### PR TITLE
ft: #184361213 add Two Factor Authentication for Sellers

### PR DESCRIPTION
**Pull request by musllim**
# Description
***Two factor authentication feauture for seller***
GIVEN there is a request to log in a user to the website WHEN 2FA is enabled on the particular user
THEN Further request for a 2FA token should be requested and once the token is received and validated, the user is allowed to log in.

add all tests for this feature plus it's documentation

Fixes # (issue)

# Type of change

Please check the options that are relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or Enhancement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have created a loom video for the new feature or enhancement 
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/controllers/twoFactorAuth.controller.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+export const create2FAToken = async (req: Request, res: Response) => {
+  try {
+    const newSecret = twofactor.generateSecret({
+      name: "Ecom-backend",
+      account: req.user.surname,
+    });
+
+    const user = await User.findByPk(req.user.id);
+
+    if (user && user?.role !== Role.Seller) {
```
Comment (emilenfc) [inline src/controllers/twoFactorAuth.controller.ts:19]: I think here you may `import { isSeller} from '../middleware/auth`'; it can do similar function

**Review by emilenfc — DISMISSED** [review]
LGTM

File: src/controllers/twoFactorAuth.controller.ts (hunk 1 above)
Comment (musllim) [reply src/controllers/twoFactorAuth.controller.ts:19]: Thanks, this saved me from hound bro

**Review by shyakadavis — DISMISSED** [review]
LGMT 🪐

**Review by emilenfc — DISMISSED** [review]

File: src/controllers/twoFactorAuth.controller.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { Request, Response } from "express";
+import User from "../database/models/User.model";
+import * as twofactor from "node-2fa";
+import sendEmailToken from "../services/mails/sendEmailToken";
+
+/*
+  this middleware will enable two factor authentication on user account, create random code and send it to user through email and have to be verified to allow user access protected routes 
+*/
```
Comment (mucyoalexaxel) [inline src/controllers/twoFactorAuth.controller.ts:8]:
```suggestion
```

**Review by SergeInGithub — DISMISSED** [review]
LGTM

File: src/controllers/twoFactorAuth.controller.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      });
+      res.send({ data: { ...newSecret, ...code }, status: 200 });
+      await sendEmailToken(user.email, code.token);
+    }
+  } catch (error) {
+    if (error instanceof Error) {
+      res.status(500).send({ error: error.message, status: 500 });
+    }
+  }
+};
```
Comment (mucyoalexaxel) [inline src/controllers/twoFactorAuth.controller.ts:34]:
```suggestion
export const create2FAToken = async (req: Request, res: Response) => {
  const { id, surname } = req.user; // Try destructing your variables at the start of your function
  try {
    const { secrete } = twofactor.generateSecret({
      name: "Ecom-backend",
      account: surname,
    });

    const user = await User.findByPk(id);

    const { token }= twofactor.generateToken(secret);

    if (user)) {
      await user.update({
        twoFAenabled: true,
        twoFAverified: false,
        secret: secret,
      });
      await sendEmailToken(user.email, token);
      res.send({ data: { secret, token }, status: 200 });
    }
    else {
              throw new Error(`User not found.`)
    }
  } catch (error) {
    if (error instanceof Error) {
      res.status(500).send({ error: error.message, status: 500 });
    }
  }
};
```

Comment (mucyoalexaxel) [reply src/controllers/twoFactorAuth.controller.ts:34]: @musllim Try refactoring this function to the proposed one, looks more clean and is easily understood. 
implement the same logic for the function `verify2FAToken`, you have some `if` conditions that don't make sense at all.

File: src/routes/auth.routes.ts [hunk 4]
```diff
@@ -10,22 +10,27 @@ import {
 } from "../controllers";
 import { ValidateJoi } from "../middleware/validation/validation.middleware";
 import { UserSchema } from "../middleware/validation/user.schema.middleware";
-import { isAuth } from "../middleware/auth";
+import { protectRoute } from "../services/protectRoutes.service";
+import { isSeller } from "../middleware/auth";
 
 const authRouter = Router();
 
+authRouter.post("/login", ValidateJoi(UserSchema.loginData.create), loginUser); // login a User
```
Comment (mucyoalexaxel) [inline src/routes/auth.routes.ts:18]:
```suggestion
authRouter.post("/login", ValidateJoi(UserSchema.loginData.create), loginUser); 
```

@musllim remove all similar comments

File: src/services/mails/sendEmailToken.ts [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import { transporter } from "../Auth.service";
+
+export default async function sendEmailToken(
+  email: string,
+  code: string,
+): Promise<void> {
+  console.log(Math.random(), "muslim");
```
Comment (mucyoalexaxel) [inline src/services/mails/sendEmailToken.ts:7]:
```suggestion
```

File: src/services/mails/sendEmailToken.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+  const mailOptions = {
+    from: '"Team Cypher" <noreply@teamcypher.com>',
+    to: email,
+    subject: "verification code",
+    text: `Hi \n\n you login verification code is ${code}!\n\n`,
+    html: `
+        
+          <p>use the following code to log in you account <strong>${code}</strong></p>
+        `,
```
Comment (mucyoalexaxel) [inline src/services/mails/sendEmailToken.ts:17]:
```suggestion
    html: ``,
```

@musllim Generate a better HTML Template

**Review by mucyoalexaxel — CHANGES_REQUESTED** [review]
Great work @musllim . Have a look at the suggested changes and refactor your code accordingly

File: src/database/models/User.model.ts [hunk 7]
```diff
@@ -100,18 +117,10 @@ const User = sequelize.define<UserInstance>(
     },
   },
   {
-    modelName: 'User',
```
[the remark this answers is not in the document]
Comment (musllim) [reply src/database/models/User.model.ts]: look on line 67 - 72

**Comment by ISHIMWE-Aime** [comment]
@musllim Hey; 
you have done great work; but try to solve conflicts on this PR

File: src/services/mails/sendEmailToken.ts (hunk 6 above)
Comment (musllim) [reply src/services/mails/sendEmailToken.ts:17]: done

File: src/controllers/twoFactorAuth.controller.ts (hunk 3 above)
Comment (musllim) [reply src/controllers/twoFactorAuth.controller.ts:34]: thanks

**Review by mucyoalexaxel — DISMISSED** [review]
LGTM

**Review by emilenfc — APPROVED** [review]

**Review by mucyoalexaxel — APPROVED** [review]
LGTM

**Pull request merged**